File: improvements/06_reporter_metabolites_ak.py
#!/usr/bin/env python3
"""
==============================================================================
Script: 06_reporter_metabolites_ak.py
Author: Dr. Ali Kaynar (King's College London)
Description: Patil & Nielsen Reporter Metabolite Algorithm with GEM Model Input (Python)
==============================================================================
"""

import os
import sys
import argparse
import numpy as np
import pandas as pd
from scipy.stats import norm

def get_root_dir():
    script_path = os.path.abspath(__file__)
    return os.path.dirname(os.path.dirname(script_path))

root_dir = get_root_dir()
out_dir = os.path.join(root_dir, "improvements", "results_06_reporter_metabolites_python")
os.makedirs(out_dir, exist_ok=True)

parser = argparse.ArgumentParser(description="GEM-based Reporter Metabolite Analysis")
parser.add_argument("--gem", type=str, default=None, help="Path to GEM model file (.xml, .csv, .yml)")
args, unknown = parser.parse_known_args()

print("============================================================")
print("AK Pipeline 06 (Python): GEM-Based Reporter Metabolite Analysis")
print("Root Directory:", root_dir)
print("Output Directory:", out_dir)
print("============================================================\n")

# 1. Resolve GEM Input Path
gem_path = args.gem
if not gem_path:
    local_xml = os.path.join(root_dir, "improvements", "data", "Human-GEM.xml")
    csv_default = os.path.join(root_dir, "improvements", "data", "human_gem_topology_network.csv")
    if os.path.exists(local_xml):
        gem_path = local_xml
    elif os.path.exists(csv_default):
        gem_path = csv_default
    else:
        gem_path = "/Users/k2254978/Desktop/Work/ToolBoxProgs/Human_GEM_2.0/model/Human-GEM.xml"

print(f"[INFO] GEM Model Input File: {gem_path}")

if gem_path.endswith(".xml"):
    from parse_gem import parse_human_gem_xml
    csv_path = os.path.join(root_dir, "improvements", "data", "human_gem_topology_network.csv")
    gem_df = parse_human_gem_xml(gem_path, csv_path)
elif gem_path.endswith(".csv"):
    gem_df = pd.read_csv(gem_path)
else:
    raise ValueError(f"Unsupported GEM model format: {gem_path}")

# 2. Currency Metabolites Filter
currency_metabolites = {"h2o", "atp", "adp", "amp", "nad+", "nadh", "nadp+", "nadph",
                        "h+", "pi", "ppi", "coa", "co2", "o2", "hco3-", "na+", "k+", "cl-", "water", "oxygen", "phosphate"}

gem_df_clean = gem_df.dropna(subset=["metabolite_name", "gene_symbol"]).copy()
gem_df_clean = gem_df_clean[~gem_df_clean["metabolite_name"].str.lower().isin(currency_metabolites)]

print(f"Loaded {len(gem_df_clean)} GEM network associations across {gem_df_clean['metabolite_name'].nunique()} metabolites.")

# 3. Load Differential Expression Data
res_file = os.path.join(root_dir, "week4-day3-nafld-gse130970-rnaseq", "results", "gse130970_results.csv")
if not os.path.exists(res_file):
    res_file = os.path.join(root_dir, "week4-day1-nafld-bulk-rnaseq", "results", "deseq2_results.csv")

def find_col(df_cols, candidates):
    for c in candidates:
        if c in df_cols:
            return c
    return None

if os.path.exists(res_file):
    print(f"[INFO] Loading differential gene expression results from: {res_file}")
    df_raw = pd.read_csv(res_file)
    cols = list(df_raw.columns)
    
    sym_col = find_col(cols, ["gene_symbol", "symbol", "Gene", "gene"])
    pval_col = find_col(cols, ["pvalue_mle", "pvalue", "p.value", "pval"])
    lfc_col = find_col(cols, ["lfc_mle", "log2FoldChange", "lfc", "lfc_apeglm"])
    
    df_clean = df_raw.dropna(subset=[sym_col, pval_col]).copy()
    df_clean["gene_symbol"] = df_clean[sym_col].astype(str)
    df_clean["pvalue"] = np.maximum(df_clean[pval_col].values, 1e-300)
    df_clean["lfc"] = df_clean[lfc_col].values if lfc_col else 0.0
    
    df_clean = df_clean.sort_values(by="pvalue").groupby("gene_symbol").first().reset_index()
else:
    print("[INFO] Generating synthetic gene dataset...")
    all_genes = list(set(gem_df_clean["gene_symbol"].tolist() + [f"GENE_{i:04d}" for i in range(1, 501)]))
    np.random.seed(42)
    df_clean = pd.DataFrame({
        "gene_symbol": all_genes,
        "pvalue": np.random.uniform(1e-5, 0.5, size=len(all_genes)),
        "lfc": np.random.normal(0, 1.5, size=len(all_genes))
    })

# 4. Calculate Gene Z-scores
df_clean["gene_z"] = norm.ppf(1.0 - df_clean["pvalue"].values / 2.0)
gene_z_dict = dict(zip(df_clean["gene_symbol"], df_clean["gene_z"]))
gene_lfc_dict = dict(zip(df_clean["gene_symbol"], df_clean["lfc"]))
gene_universe = list(gene_z_dict.keys())

# 5. Patil & Nielsen Reporter Algorithm across GEM Nodes
met_groups = gem_df_clean.groupby("metabolite_name")["gene_symbol"].unique().to_dict()
qual_mets = {m: genes for m, genes in met_groups.items() if 3 <= len([g for g in genes if g in gene_z_dict]) <= 100}

print(f"Running Reporter Metabolite Analysis across {len(qual_mets)} qualified GEM nodes...")

n_perm = 1000
results = []

for idx, (met, genes) in enumerate(qual_mets.items()):
    g_set = [g for g in genes if g in gene_z_dict]
    k = len(g_set)
    
    z_scores = [gene_z_dict[g] for g in g_set]
    lfcs = [gene_lfc_dict[g] for g in g_set]
    
    z_raw = np.sum(z_scores) / np.sqrt(k)
    avg_lfc = np.mean(lfcs)
    
    np.random.seed(42 + idx)
    perm_z = [np.sum(np.random.choice(list(gene_z_dict.values()), size=k, replace=False)) / np.sqrt(k) for _ in range(n_perm)]
    
    mu_k = np.mean(perm_z)
    sigma_k = np.std(perm_z)
    z_corr = (z_raw - mu_k) / sigma_k if sigma_k > 0 else z_raw
    p_rep = 1.0 - norm.cdf(z_corr)
    
    results.append({
        "Metabolite": met,
        "Neighbor_Genes_Count": k,
        "Mean_Log2FC": round(avg_lfc, 3),
        "Z_Raw": round(z_raw, 3),
        "Reporter_Z_Score": round(z_corr, 3),
        "Pvalue": float(f"{p_rep:.4e}"),
        "Neighbor_Genes": "; ".join(g_set[:8])
    })

res_df = pd.DataFrame(results).sort_values(by="Reporter_Z_Score", ascending=False)
print(res_df.head(15).to_string(index=False))

# 6. Save CSV & PNG Plot
out_csv = os.path.join(out_dir, "gem_reporter_metabolites_python.csv")
res_df.to_csv(out_csv, index=False)
print(f"\nSaved CSV: {out_csv}")

try:
    import matplotlib.pyplot as plt
    
    top_df = res_df.head(20)
    plt.figure(figsize=(8, 6.5))
    colors = ["#D81B60" if lfc > 0 else "#1E88E5" for lfc in top_df["Mean_Log2FC"]]
    plt.barh(top_df["Metabolite"], top_df["Reporter_Z_Score"], color=colors, edgecolor="black")
    plt.xlabel("Corrected Reporter Z-score", fontsize=11, fontweight="bold")
    plt.ylabel("Metabolite Node", fontsize=11, fontweight="bold")
    plt.title("Top 20 GEM Reporter Metabolites (Human-GEM Input - Python)", fontsize=12, fontweight="bold")
    plt.gca().invert_yaxis()
    plt.tight_layout()
    
    out_png = os.path.join(out_dir, "gem_reporter_metabolites_python.png")
    plt.savefig(out_png, dpi=300)
    plt.close()
    print(f"Saved PNG Plot: {out_png}")
except Exception as e:
    print(f"[NOTE] Matplotlib plot skipped: {e}")

print("[SUCCESS] Python Pipeline 06 completed successfully using GEM model input!")

File: improvements/06_reporter_metabolites_ak.R
#!/usr/bin/env Rscript
## ==============================================================================
## Script: 06_reporter_metabolites_ak.R
## Author: Dr. Ali Kaynar (King's College London)
## Description: Patil & Nielsen Reporter Metabolite Pipeline with GEM Model Input
## ==============================================================================

suppressPackageStartupMessages({
  library(dplyr)
  library(ggplot2)
})

# Environment & Path Auto-Detection
get_root_dir <- function() {
  cmd_args <- commandArgs(trailingOnly = FALSE)
  file_arg <- grep("--file=", cmd_args, value = TRUE)
  if (length(file_arg) > 0) {
    return(dirname(dirname(normalizePath(sub("--file=", "", file_arg[1])))))
  }
  if (exists(".rs.getScriptPath", mode = "function")) {
    sp <- .rs.getScriptPath()
    if (!is.null(sp) && nchar(sp) > 0) return(dirname(dirname(normalizePath(sp))))
  }
  return(getwd())
}

root_dir <- get_root_dir()
out_dir  <- file.path(root_dir, "improvements", "results_06_reporter_metabolites")
if (!dir.exists(out_dir)) dir.create(out_dir, recursive = TRUE)

cat("============================================================\n")
cat("AK Pipeline 06: GEM-Based Reporter Metabolite Analysis\n")
cat("Root Directory:", root_dir, "\n")
cat("Output Directory:", out_dir, "\n")
cat("============================================================\n\n")

# 1. Accept GEM Model File Input or Default to Extracted Human-GEM Network
args <- commandArgs(trailingOnly = TRUE)
gem_input_file <- if (length(args) > 0) args[1] else file.path(root_dir, "improvements", "data", "human_gem_topology_network.csv")

if (!file.exists(gem_input_file)) {
  # Fallback workspace paths
  alt_gem <- "/Users/k2254978/Desktop/Work/ToolBoxProgs/Human_GEM_2.0/model/Human-GEM.xml"
  if (file.exists(alt_gem)) {
    cat("[INFO] Running GEM XML Parser on:", alt_gem, "\n")
    system(sprintf("python3 %s/improvements/parse_gem.py", root_dir))
  }
  gem_input_file <- file.path(root_dir, "improvements", "data", "human_gem_topology_network.csv")
}

cat("[INFO] Loading GEM Model Topology Input from:", gem_input_file, "\n")

if (file.exists(gem_input_file)) {
  gem_map <- read.csv(gem_input_file, stringsAsFactors = FALSE)
} else {
  stop("GEM model input file not found.")
}

# 2. Currency Metabolites & Noise Filtering
currency_metabolites <- c("H2O", "ATP", "ADP", "AMP", "NAD+", "NADH", "NADP+", "NADPH",
                          "H+", "Pi", "PPi", "CoA", "CO2", "O2", "HCO3-", "Na+", "K+", "Cl-", "water", "oxygen", "phosphate")

gem_map_clean <- gem_map %>%
  filter(!is.na(metabolite_name), metabolite_name != "") %>%
  filter(!is.na(gene_symbol), gene_symbol != "") %>%
  filter(!tolower(metabolite_name) %in% tolower(currency_metabolites))

cat(sprintf("GEM Topology: Extracted %d metabolite-gene associations across %d unique metabolites.\n",
            nrow(gem_map_clean), length(unique(gem_map_clean$metabolite_name))))

# 3. Load DESeq2 Results (Real or Synthetic Fallback)
res_file <- file.path(root_dir, "week4-day3-nafld-gse130970-rnaseq", "results", "gse130970_results.csv")
if (!file.exists(res_file)) {
  res_file <- file.path(root_dir, "week4-day1-nafld-bulk-rnaseq", "results", "deseq2_results.csv")
}

find_col <- function(cols, candidates) {
  match <- intersect(candidates, cols)
  if (length(match) > 0) return(match[1])
  return(NULL)
}

if (file.exists(res_file)) {
  cat("[INFO] Loading differential gene expression results from:", res_file, "\n")
  df_raw <- read.csv(res_file, stringsAsFactors = FALSE)
  cols   <- colnames(df_raw)
  
  sym_col  <- find_col(cols, c("gene_symbol", "symbol", "Gene", "gene"))
  pval_col <- find_col(cols, c("pvalue_mle", "pvalue", "p.value", "pval"))
  lfc_col  <- find_col(cols, c("lfc_mle", "log2FoldChange", "lfc", "lfc_apeglm"))
  
  df_clean <- df_raw %>%
    filter(!is.na(.data[[sym_col]]), .data[[sym_col]] != "") %>%
    filter(!is.na(.data[[pval_col]])) %>%
    mutate(
      gene_symbol = .data[[sym_col]],
      pvalue = pmax(.data[[pval_col]], 1e-300),
      lfc = if (!is.null(lfc_col)) .data[[lfc_col]] else 0
    ) %>%
    group_by(gene_symbol) %>%
    slice_min(order_by = pvalue, n = 1, with_ties = FALSE) %>%
    ungroup()
} else {
  cat("[INFO] Generating synthetic gene p-value vector...\n")
  set.seed(77)
  all_genes <- unique(c(gem_map_clean$gene_symbol, paste0("GENE_", sprintf("%04d", 1:500))))
  df_clean <- data.frame(
    gene_symbol = all_genes,
    pvalue = runif(length(all_genes), 1e-5, 0.5),
    lfc = rnorm(length(all_genes), 0, 1.5),
    stringsAsFactors = FALSE
  )
}

# 4. Calculate Two-Tailed Gene Z-scores
df_clean <- df_clean %>%
  mutate(gene_z = qnorm(1 - pvalue / 2))

gene_z_vec   <- setNames(df_clean$gene_z, df_clean$gene_symbol)
gene_lfc_vec <- setNames(df_clean$lfc, df_clean$gene_symbol)
gene_universe <- names(gene_z_vec)

# 5. Patil & Nielsen Algorithm on GEM Nodes
# Filter metabolites to those with 3 to 100 neighboring genes in the dataset
metabolite_gene_counts <- gem_map_clean %>%
  filter(gene_symbol %in% gene_universe) %>%
  group_by(metabolite_name) %>%
  summarise(
    n_genes = n_distinct(gene_symbol),
    genes = list(unique(gene_symbol)),
    .groups = "drop"
  ) %>%
  filter(n_genes >= 3, n_genes <= 100)

cat(sprintf("Running Reporter Metabolites on %d qualified GEM nodes (3 <= k <= 100)...\n", nrow(metabolite_gene_counts)))

n_perm <- 1000

reporter_results <- lapply(1:nrow(metabolite_gene_counts), function(i) {
  met <- metabolite_gene_counts$metabolite_name[i]
  g_set <- metabolite_gene_counts$genes[[i]]
  k <- length(g_set)
  
  z_raw <- sum(gene_z_vec[g_set], na.rm = TRUE) / sqrt(k)
  avg_lfc <- mean(gene_lfc_vec[g_set], na.rm = TRUE)
  
  # Background Monte Carlo Permutations
  set.seed(42 + i)
  perm_z <- replicate(n_perm, {
    rand_g <- sample(gene_universe, k)
    sum(gene_z_vec[rand_g], na.rm = TRUE) / sqrt(k)
  })
  
  mu_k    <- mean(perm_z)
  sigma_k <- sd(perm_z)
  z_corr  <- if (sigma_k > 0) (z_raw - mu_k) / sigma_k else z_raw
  p_rep   <- pnorm(z_corr, lower.tail = FALSE)
  
  data.frame(
    Metabolite = met,
    Neighbor_Genes_Count = k,
    Mean_Log2FC = round(avg_lfc, 3),
    Z_Raw = round(z_raw, 3),
    Reporter_Z_Score = round(z_corr, 3),
    Pvalue = signif(p_rep, 4),
    Neighbor_Genes = paste(head(g_set, 8), collapse = "; "),
    stringsAsFactors = FALSE
  )
}) %>% bind_rows() %>%
  mutate(Padj = signif(p.adjust(Pvalue, method = "BH"), 4)) %>%
  arrange(desc(Reporter_Z_Score))

print(head(reporter_results, 15))

# 6. Export Results & Plots
write.csv(reporter_results, file.path(out_dir, "gem_reporter_metabolites_summary.csv"), row.names = FALSE)

top_plot_df <- head(reporter_results, 20)

p_rep <- ggplot(top_plot_df, aes(x = reorder(Metabolite, Reporter_Z_Score), y = Reporter_Z_Score, fill = Mean_Log2FC)) +
  geom_bar(stat = "identity", width = 0.7, color = "black") +
  coord_flip() +
  scale_fill_gradient2(low = "#1E88E5", mid = "grey90", high = "#D81B60", midpoint = 0, name = "Mean Log2FC") +
  theme_bw(base_size = 11) +
  labs(
    title = "Top 20 GEM Reporter Metabolites (Human-GEM 2.0 Input)",
    subtitle = "Patil & Nielsen Z-score correction on GEM stoichiometric topology",
    x = "Metabolite Node",
    y = "Corrected Reporter Z-score"
  ) +
  theme(plot.title = element_text(face = "bold"))

ggsave(file.path(out_dir, "gem_reporter_metabolites_zscores.png"), p_rep, width = 8, height = 6.5, dpi = 300)
cat("[SUCCESS] Pipeline 06 completed using GEM input. Results saved to:", out_dir, "\n")
